apache/metron
# METRON-1301 Sorting on Triage Score Unexpectedly Filters Some Recods

UPDATE:  To help avoid causing too much pain with merge conflicts, I'd like to see the following PRs resolved before we consider merging this.
- [x]  #824
- [x]  #825
- [x]  #827

## Changes

* Adds logic to sort 'missing' fields (fields that are not present in all of the underlying indices.)  For example, threat triage score is only present in the records that have been triaged.  Without this PR, sorting on the threat triage score would unexpectedly filter the search results.
    * When sorted descending, missing fields are sorted last.
    * When sorted ascending, missing fields are sorted first.

* Adds validation and error checking of the response received from Elasticsearch.  Previously there was little to no validation of the response.  This also handles partial shard failures; errors that occur only on a subset of the data.  Without this change some errors that occurred within Elasticsearch were not visible.

* Updates to the index templates ensure that the threat triage fields like threat triage overall score and the threat triage rules are mapped to a correct type.  These fields were not mapped correctly.

* Adds additional logging in the ElasticsearchDAO so that errors can be traced and debugged.

* Refactored the ElasticsearchDAO and extracted some logic into two new interfaces.  This was done so that I could write tests for the new logic.  It was not possible to specifically test this change without refactoring due to limitations with mocking.
    * The column metadata logic is now contained in `ColumnMetadataDao` and `ElasticsearchColumnMetadataDao`.  Without pulling this logic out, it was not possible to test the new sorting logic.
    * The logic for submitting a request to Elasticsearch and validating the response is now contained within `ElasticsearchSearchSubmitter` class.  Without refactoring this it was not possible to test the  new response handling logic.

## Testing

1. Launch Full Dev
2. Allow some alerts to accumulate so that when you open the Alerts UI there are a fair number of alerts to work with.
3. By default it will sort by timestamp.  Take note of the number of records returned.  
4. Click on the "Score" field to sort the results by threat triage score.
5. Note the total number of records returned after sorting.  This should be equal to or greater than what you saw before.
6. The highest scores should be sorted first.  Records that do not contain a threat triage score should be sorted last.
7. Click on the "Score" field to reverse the sort order.  Records without a threat triage score should now be sorted first.

## Pull Request Checklist
